parlai/core/utils.py

```diff
@@ -713,6 +715,52 @@ def __contains__(self, key):
         """Determine if text contains any offensive words in the filter."""
         return self.contains_offensive_language(key)
 
+    def str_segment(self, text, dict_agent, max_length):
+        '''
+        Function that segments a word without spaces into the most
+        probable phrase with spaces
+
+        :param string text: string to segment
+        :param DictionaryAgent dict_agent: Dictionary we use
+            to look at word frequencies
+        :param int max_length: max_length of string to segment
+
+        '''
+        freqs = dict_agent.freqs()
+
+        # Total number of word tokensd
+        N = sum(freqs.values())
+
+        # Number of distinct words in the Vocab
+        V = len(freqs)
+
+        logNV = math.log(N + V)
+
+        @lru_cache(maxsize = None)
+        def segment(text):
+            # Return a list of words that is the best segmentation of text.
+            if not text:
+                return []
+            candidates = ([first] + segment(rem) for first, rem in splits(text, max_length))
+            return max(candidates, key = prob_words)
+
+        def splits(text, max_length):
+            # Returns a list of all possible first and remainder tuples where
+            return [(text[:i+1], text[i+1:])
+                for i in range(min(len(text), max_length))]
+
+        def prob_words(words):
```

Comment:
this isn't quite the probability, it's really a transformation of the log probability. maybe name it `score` or something.